torchdynamo: add convolution add(relu) inplace fusion kernel

XiaobingSuper opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88794
* #88414
* #88412
* #88403
* __->__ #88048

This PR is about add convolution add(relu) inplace fusion kernel which  works for **other.add_(conv)**.


cc @VitalyFedyunin @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @Guobing-Chen @chunyuan-w @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

XiaobingSuper committed:
torchdynamo: add convolution add(relu) inplace fusion kernel

[ghstack-poisoned]

XiaobingSuper committed:
Update on "torchdynamo: add convolution add(relu) inplace fusion kernel"




[ghstack-poisoned]

XiaobingSuper committed:
Update on "torchdynamo: add convolution add(relu) inplace fusion kernel"




cc VitalyFedyunin jgong5 mingfeima sanchitintel ashokei jingxu10

[ghstack-poisoned]

XiaobingSuper committed:
Update on "torchdynamo: add convolution add(relu) inplace fusion kernel"




cc VitalyFedyunin jgong5 mingfeima sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -416,32 +444,53 @@ Tensor mkldnn_convolution_pointwise_binary(
       "mkldnn_convolution_add_: currently only support 2d and 3d")
+  TORCH_CHECK(
+      binary_attr == "add",
+      "mkldnn_convolution_pointwise_binary_: currently only add inplace fusion")
```
```suggestion
      "mkldnn_convolution_pointwise_binary_: only support add binary op fusion")
```

jgong5 commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -416,32 +444,53 @@ Tensor mkldnn_convolution_pointwise_binary(
+      "mkldnn_convolution_pointwise_binary_: currently only add inplace fusion")
+  TORCH_CHECK(
+      !alpha.has_value() || alpha.value().to<float>() == 1.0,
+      "mkldnn_convolution_pointwise_binary: the alpha value should be none or 1.0");
```
```suggestion
      "mkldnn_convolution_pointwise_binary: the alpha value for the binary op should be none (meaning 1.0) or 1.0");
```

jgong5 commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -416,32 +444,53 @@ Tensor mkldnn_convolution_pointwise_binary(
+      "mkldnn_convolution_pointwise_binary: the alpha value should be none or 1.0");
+  TORCH_CHECK(
+      !unary_attr.has_value() || unary_attr.value() == "relu",
+      "mkldnn_convolution_pointwise_binary: the unary_attr value should be none or relu");
```
```suggestion
      "mkldnn_convolution_pointwise_binary: only support none or relu unary op fusion after binary op");
```

jgong5 commented on aten/src/ATen/native/mkldnn/Utils.cpp:
```diff
@@ -140,6 +140,13 @@ const std::map<c10::string_view, AttrFunction>& fx_fusion_attr_map() {
   return fusion_attr_map;
 };
 
+const std::map<c10::string_view, ideep::algorithm>& fusion_unary_alg_map() {
+  static const std::map<c10::string_view, ideep::algorithm> fusion_attr_map{
+      {"relu", {ideep::algorithm::eltwise_relu}},
+  };
+  return fusion_attr_map;
+};
+
```
Why can't we reuse `fx_fusion_attr_map`? Also, should `fx_fusion_attr_map` be renamed to `fusion_unary_attr_map`?

XiaobingSuper replied:
if we reuse **fx_fusion_attr_map**, we need to query the eltwise alg and its paraemters to re-create a new op attr, so I think we can have an alg map as binary and directly query the eltwise alg. I have changed the  **fx_fusion_attr_map** to **fusion_unary_attr_map**.

XiaobingSuper committed:
Update on "torchdynamo: add convolution add(relu) inplace fusion kernel"




cc VitalyFedyunin jgong5 mingfeima sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

XiaobingSuper committed:
Update on "torchdynamo: add convolution add(relu) inplace fusion kernel"




cc VitalyFedyunin jgong5 mingfeima sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

XiaobingSuper commented on aten/src/ATen/native/mkldnn/Conv.cpp:
```diff
@@ -322,10 +322,17 @@ Tensor mkldnn_convolution_pointwise_binary(
+    c10::optional<at::Scalar> alpha,
+    c10::optional<c10::string_view> unary_attr,
+    torch::List<c10::optional<at::Scalar>> unary_scalars,
+    c10::optional<c10::string_view> unary_algorithm) {
```
some explanations are added.

XiaobingSuper committed:
Update on "torchdynamo: add convolution add(relu) inplace fusion kernel"




cc VitalyFedyunin jgong5 mingfeima sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

XiaobingSuper commented on test/test_mkldnn_fusion.py:
```diff
@@ -282,13 +282,53 @@ def forward(self, x, other):
                         )
                     self.assertEqual(ref, fused)
 
+    def test_conv_add_relu_inplace(self):
```
yes, I changed it by add this test case into test_conv_binary_fusion_ops.

XiaobingSuper committed:
Update on "torchdynamo: add convolution add(relu) inplace fusion kernel"


This PR is about add convolution add(relu) inplace fusion kernel which  works for **other.add_(conv)**.


cc VitalyFedyunin jgong5 mingfeima sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge

XiaobingSuper committed:
Update on "torchdynamo: add convolution add(relu) inplace fusion kernel"


This PR is about add convolution add(relu) inplace fusion kernel which  works for **other.add_(conv)**.


cc VitalyFedyunin jgong5 mingfeima sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper committed:
Update on "torchdynamo: add convolution add(relu) inplace fusion kernel"


This PR is about add convolution add(relu) inplace fusion kernel which  works for **other.add_(conv)**.


cc VitalyFedyunin jgong5 mingfeima sanchitintel ashokei jingxu10 mlazos soumith voznesenskym yanboliang penguinwu anijain2305 EikanWang Guobing-Chen chunyuan-w zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 desertfire

[ghstack-poisoned]

XiaobingSuper commented:
@pytorchbot merge
